agent_holder/src/offer/aggregate.rs

```diff
@@ -0,0 +1,485 @@
+    // TODO: These should not be part of this Aggregate. Instead, an Event Subscriber should be listening to the
+    // `CredentialResponseReceived` event and then trigger the `CredentialCommand::AddCredential` command. We can do
+    // this once we have a mechanism implemented that can both listen to events as well as trigger commands.
+    pub credentials: Vec<serde_json::Value>,
+}
+
+#[async_trait]
+impl Aggregate for Offer {
+    type Command = OfferCommand;
+    type Event = OfferEvent;
+    type Error = OfferError;
+    type Services = Arc<HolderServices>;
+
+    fn aggregate_type() -> String {
+        "offer".to_string()
+    }
+
+    async fn handle(&self, command: Self::Command, services: &Self::Services) -> Result<Vec<Self::Event>, Self::Error> {
+        use OfferCommand::*;
+        use OfferError::*;
+        use OfferEvent::*;
+
+        info!("Handling command: {:?}", command);
+
+        match command {
+            ReceiveCredentialOffer {
+                offer_id,
+                credential_offer,
+            } => {
+                let wallet = &services.wallet;
+
+                let credential_offer = match credential_offer {
+                    CredentialOffer::CredentialOfferUri(credential_offer_uri) => services
+                        .wallet
+                        .get_credential_offer(credential_offer_uri)
+                        .await
+                        .map_err(|_| CredentialOfferByReferenceRetrievalError)?,
+                    CredentialOffer::CredentialOffer(credential_offer) => *credential_offer,
+                };
+
+                // The credential offer contains a credential issuer url.
+                let credential_issuer_url = credential_offer.credential_issuer.clone();
+
+                // Get the credential issuer metadata.
+                let credential_issuer_metadata = wallet
+                    .get_credential_issuer_metadata(credential_issuer_url.clone())
+                    .await
+                    .map_err(|_| CredentialIssuerMetadataRetrievalError)?;
+
+                let credential_configurations: HashMap<String, CredentialConfigurationsSupportedObject> =
+                    credential_issuer_metadata
+                        .credential_configurations_supported
+                        .iter()
+                        .filter(|(id, _)| credential_offer.credential_configuration_ids.contains(id))
+                        .map(|(id, credential_configuration)| (id.clone(), credential_configuration.clone()))
+                        .collect();
+
+                Ok(vec![CredentialOfferReceived {
+                    offer_id,
+                    credential_offer: Box::new(credential_offer),
+                    credential_configurations,
+                }])
+            }
+            AcceptCredentialOffer { offer_id } => {
+                // TODO: should we 'do nothing' or log a `warn!` message instead of returning an error?
```

Comment:
If this cannot happen from a business logic perspective, it's a "logical error" and should be logged and treated as any other error. Can we recover from it? Delete the offer? Auto-reject it?

Reply:
We agreed to process the offer again, even if it is not in the "pending" state. A user can accept an offer multiple times. A warning is still always logged. If there is no actual business case for accepting an offer multiple times, this part of the code can throw an error instead and abort the issuance.